src/main/java/servlets/ReceiptAnalysis.java

```diff
@@ -53,6 +66,29 @@ private static ByteString readImageBytes(String url) throws IOException {
     return imgBytes;
   }
 
+  /** Retrieves the binary data stored at the given blob key. */
+  private static ByteString readImageBytes(BlobKey blobKey) throws IOException {
+    BlobstoreService blobstoreService = BlobstoreServiceFactory.getBlobstoreService();
+    BlobInfo blobInfo = new BlobInfoFactory().loadBlobInfo(blobKey);
+    long blobSize = blobInfo.getSize();
+
+    ByteArrayOutputStream outputBytes = new ByteArrayOutputStream();
+    int fetchSize = BlobstoreService.MAX_BLOB_FETCH_SIZE;
+    long currentByteIndex = 0;
+
+    // Fetch all the bytes from the blob in fragments of the maximum fetch size.
+    while (currentByteIndex < blobSize) {
+      // End index is inclusive, so subtract 1 to get fetchSize bytes.
+      byte[] bytes =
+          blobstoreService.fetchData(blobKey, currentByteIndex, currentByteIndex + fetchSize - 1);
```

Comment:
I assume it's okay for the end index to be greater than the blobSize?

Reply:
Yes, fetchData() will not read past the blob with the specified blob key.